app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,29 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def self.provides_callback_for(provider)
+    class_eval %{
+      def #{provider}
+        @user = User.find_for_oauth(env["omniauth.auth"], current_user)
+
+        if @user.persisted?
+          sign_in_and_redirect @user, event: :authentication
+          set_flash_message(:notice, :success, kind: "#{provider}".capitalize) if is_navigational_format?
+        else
+          session["devise.#{provider}_data"] = env["omniauth.auth"]
+          redirect_to new_user_registration_url
+        end
+      end
+    }
```

Comment:
What is this crazy nonsense? It would probably be clearer to explicitly define the callbacks for Google and Facebook since you only have two of them.